File: backend/internal/service/reports_04.go
package service


import (
	"context"
	"time"

	"github.com/biruk-ak/clinical-lis/backend/internal/models"
	"github.com/biruk-ak/clinical-lis/backend/internal/repository"
	"github.com/google/uuid"
)

// ReportRecord04Service encapsulates business rules for ReportRecord04.
type ReportRecord04Service struct {
	repo *repository.ReportRecord04Repository
}

// NewReportRecord04Service wires the service.
func NewReportRecord04Service(repo *repository.ReportRecord04Repository) *ReportRecord04Service {
	return &ReportRecord04Service{repo: repo}
}

// Create validates and persists a new ReportRecord04.
func (s *ReportRecord04Service) Create(ctx context.Context, actor, facilityID, code, display, status string, priority int, notes string) (*models.ReportRecord04, error) {
	now := time.Now().UTC()
	rec := &models.ReportRecord04{
		ID:           uuid.NewString(),
		ExternalCode: code,
		DisplayName:  display,
		Status:       status,
		Priority:     priority,
		FacilityID:   facilityID,
		CreatedBy:    actor,
		UpdatedBy:    actor,
		Notes:        notes,
		MetadataJSON: "{}",
		Version:      1,
		IsActive:     true,
		CreatedAt:    now,
		UpdatedAt:    now,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	if err := s.repo.Create(ctx, rec); err != nil {
		return nil, err
	}
	return rec, nil
}

// Get returns a ReportRecord04 by id.
func (s *ReportRecord04Service) Get(ctx context.Context, id string) (*models.ReportRecord04, error) {
	return s.repo.GetByID(ctx, id)
}

// List returns facility-scoped ReportRecord04 pages.
func (s *ReportRecord04Service) List(ctx context.Context, facilityID string, limit, offset int) ([]models.ReportRecord04, error) {
	if limit <= 0 || limit > 200 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}
	return s.repo.List(ctx, facilityID, limit, offset)
}

// Update mutates an existing ReportRecord04.
func (s *ReportRecord04Service) Update(ctx context.Context, actor string, rec *models.ReportRecord04) error {
	if err := rec.Validate(); err != nil {
		return err
	}
	rec.Touch(actor, time.Now().UTC())
	return s.repo.Update(ctx, rec)
}

// Archive soft-deletes a ReportRecord04.
func (s *ReportRecord04Service) Archive(ctx context.Context, id, actor string) error {
	return s.repo.Delete(ctx, id, actor)
}

// Search finds ReportRecord04 by external code prefix.
func (s *ReportRecord04Service) Search(ctx context.Context, facilityID, prefix string) ([]models.ReportRecord04, error) {
	return s.repo.SearchByCode(ctx, facilityID, prefix, 100)
}

// Stats returns active counts for dashboards.
func (s *ReportRecord04Service) Stats(ctx context.Context, facilityID string) (int64, error) {
	return s.repo.CountActive(ctx, facilityID)
}

File: frontend/src/types/reports04.ts
/** Domain types for reports / ReportRecord04 */
export type ReportRecord04Status = 'draft' | 'active' | 'pending' | 'completed' | 'cancelled' | 'archived';

export interface ReportRecord04 {
  id: string;
  externalCode: string;
  displayName: string;
  status: ReportRecord04Status;
  priority: number;
  facilityId: string;
  createdBy: string;
  updatedBy: string;
  notes: string;
  metadataJson: string;
  version: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string | null;
}

export interface ReportRecord04CreateRequest {
  externalCode: string;
  displayName: string;
  status: ReportRecord04Status;
  priority: number;
  facilityId: string;
  notes?: string;
}

export interface ReportRecord04ListResponse {
  items: ReportRecord04[];
  total?: number;
}

export interface ReportRecord04StatsResponse {
  activeCount: number;
}

export function isReportRecord04Active(rec: ReportRecord04): boolean {
  return rec.isActive && rec.status !== 'archived' && rec.status !== 'cancelled';
}

export function formatReportRecord04Label(rec: ReportRecord04): string {
  return `${rec.externalCode} — ${rec.displayName}`;
}

export function compareReportRecord04ByPriority(a: ReportRecord04, b: ReportRecord04): number {
  if (a.priority !== b.priority) return b.priority - a.priority;
  return a.displayName.localeCompare(b.displayName);
}
